### PR TITLE
feat: add Terminal-Bench benchmark integration

### LINKED ISSUE #48: benchmark: terminal_bench
Integrate terminal_bench to v1 benchmarks.

See
- Terminal-Bench https://github.com/laude-institute/terminal-bench
- OpenHands v0 integration https://github.com/All-Hands-AI/OpenHands/tree/main/evaluation/benchmarks/terminal_bench

### PR DESCRIPTION
**Pull request by neubig**
## Summary

This PR integrates Terminal-Bench evaluation into the benchmarks repository using Harbor with the openhands-sdk agent.

Fixes #48

### New Module: `benchmarks/terminalbench/`

- **`run_infer.py`**: Runs Terminal-Bench evaluation via Harbor CLI with openhands-sdk agent
  - Supports full dataset evaluation or specific task selection
  - Converts ATIF trajectories to standard JSONL format
  - Configurable timeouts and parallelism

- **`eval_infer.py`**: Processes results and generates evaluation reports
  - Aggregates metrics (cost, tokens, pass rate)
  - Produces SWE-Bench-style report JSON

- **`config.py`**: Default configuration for Terminal-Bench evaluation

- **`README.md`**: Documentation with usage examples

### CLI Entrypoints

- `terminalbench-infer`: Run Terminal-Bench inference
- `terminalbench-eval`: Process and report results

### Tests

Added comprehensive tests covering:
- Result processing and aggregation
- Metrics calculation
- Error handling
- ATIF trajectory parsing

## Usage

```bash
# Run full evaluation
uv run terminalbench-infer .llm_config/claude.json

# Run specific tasks
uv run terminalbench-infer .llm_config/claude.json --task-id hello-world

# Evaluate results
uv run terminalbench-eval ./evaluation_outputs/.../output.jsonl
```

## Architecture

The integration uses:
- **Harbor**: Official Terminal-Bench harness for task management
- **openhands-sdk agent**: Runs inside task containers (merged in [harbor#738](https://github.com/laude-institute/harbor/pull/738))
- **ATIF format**: Agent Trajectory Interchange Format for results

## Testing

All tests pass:
```
tests/test_terminalbench.py: 9 passed
```

Pre-commit checks pass:
- Ruff format ✅
- Ruff lint ✅
- PEP8 style ✅
- Pyright type check ✅

@neubig can click here to [continue refining the PR](https://app.all-hands.dev/conversations/None)

---

### ✅ Full E2E Run Verified (3 tasks, real LLM calls, verified results)

Ran 3 Terminal-Bench sample tasks with `claude-sonnet-4-20250514` via the LiteLLM proxy. All 3 completed with real agent execution — no auth errors, no exceptions, and verifier results are confirmed correct.

**1. `terminalbench-infer` — Harbor runs agent in Docker containers:**
```
$ uv run terminalbench-infer llm_config.json \
    --dataset terminal-bench-sample --num-workers 1 \
    --task-id chess-best-move --task-id polyglot-c-py --task-id sqlite-with-gcov

INFO  Harbor evaluation completed successfully
INFO  stdout: 3/3 Mean: 0.000
      Trials: 3 | Errors: 0 | Mean: 0.000
INFO  Found 3 trial results
INFO  Processed trial chess-best-move: reward=0.0
INFO  Processed trial polyglot-c-py: reward=0.0
INFO  Processed trial sqlite-with-gcov: reward=0.0
INFO  Wrote 3 successful + 0 failed entries to .../output.jsonl
```

**2. Agent logs confirm real LLM calls:**

| Task | Input Tokens | Output Tokens | Cost | Reward |
|------|-------------|---------------|------|--------|
| chess-best-move | 63,364 | 2,322 | $0.09 | 0.0 |
| polyglot-c-py | 2,247,246 | 16,692 | $1.05 | 0.0 |
| sqlite-with-gcov | 637,198 | 4,346 | $0.34 | 0.0 |

**3. Verifier results confirmed correct (all are genuine agent failures):**

- **chess-best-move**: Agent wrote `d2d4` but the correct checkmate-in-one moves are `g2g4` and `e2e4`. Wrong move → `reward=0.0` ✅
- **polyglot-c-py**: Agent created a working polyglot file but left a compiled `cmain` binary in `/app/polyglot/`. Verifier expected only `main.py.c` in that directory → `reward=0.0` ✅
- **sqlite-with-gcov**: Agent compiled SQLite with gcov to `/app/sqlite/` but didn't add it to PATH. Verifier runs `sqlite3` directly → `FileNotFoundError` → `reward=0.0` ✅

**4. `terminalbench-eval` processes results correctly:**
```
$ uv run terminalbench-eval .../output.jsonl
INFO  Total instances: 3
INFO  Completed instances: 3
INFO  Resolved instances: 0
INFO  Error instances: 0
INFO  Success rate: 0.0%
INFO  Total cost: $1.4845
```

**5. All 15 unit tests pass:**
```
$ uv run pytest tests/test_terminalbench.py -v
TestProcessTerminalbenchResults (8 tests) — all PASSED
TestConvertHarborToEvalOutput (7 tests) — all PASSED
15 passed
```

**6. Pre-commit checks pass:**
- Ruff format ✅ | Ruff lint ✅ | PEP8 style ✅ | Pyright type check ✅

### Changes from review (commit 54a3d53)

- Fixed harbor CLI flags: `--output-dir` → `--jobs-dir`, `--max-workers` → `--n-concurrent`, `--task-id` → `--task-name`
- Pass LLM credentials via `--ae` (agent env) flags instead of subprocess env
- Use absolute path for `--jobs-dir` to avoid CWD-relative path issues
- Removed unused `--timeout` parameter
- Rewrote `convert_harbor_to_eval_output` for actual harbor output structure (`TIMESTAMP/TRIAL_NAME/result.json`)
- Updated all tests to match new harbor output structure

### CODE REVIEW

File: benchmarks/terminalbench/eval_infer.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    """
+    logger.info(f"Processing {input_file} to generate report: {output_file}")
+
+    completed_ids: list[str] = []
+    resolved_ids: list[str] = []
+    unresolved_ids: list[str] = []
+    incomplete_ids: list[str] = []
+    error_ids: list[str] = []
+
+    completed_seen: set[str] = set()
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/eval_infer.py]: Fixed in 5d4fed6 - refactored to use only sets for O(1) lookup and deduplication. Lists are only created at report generation time via sorted().

File: benchmarks/terminalbench/run_infer.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    # Look for trajectory files in harbor output
+    # Harbor typically stores trajectories in trials/*/trajectory.json
+    trajectory_patterns = [
+        harbor_output_dir / "trials" / "*" / "trajectory.json",
+        harbor_output_dir / "*" / "trajectory.json",
+        harbor_output_dir / "trajectory.json",
+    ]
+
+    trajectory_files: list[Path] = []
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/run_infer.py]: Fixed in 5d4fed6 - simplified to use canonical trials/*/trajectory.json location with fail-fast if directory structure is unexpected.

File: benchmarks/terminalbench/run_infer.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+                "history": history,
+                "metrics": {
+                    "total_prompt_tokens": final_metrics.get("total_prompt_tokens", 0),
+                    "total_completion_tokens": final_metrics.get(
+                        "total_completion_tokens", 0
+                    ),
+                    "total_cost_usd": final_metrics.get("total_cost_usd", 0.0),
+                },
+            }
+            results.append(eval_entry)
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/run_infer.py:231]: Fixed in 5d4fed6 - now records errors per trajectory and raises RuntimeError if all trajectories fail. Error entries are written to output for visibility.

File: benchmarks/terminalbench/config.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    # Default timeout for agent execution (in seconds)
+    "timeout": 3600,  # 1 hour
+}
+
+# Terminal-Bench task categories (for reference)
+TASK_CATEGORIES = [
+    "hello-world",
+    "basic",
+    "intermediate",
+    "advanced",
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/config.py]: Fixed in 5d4fed6 - removed unused TASK_CATEGORIES.

File: tests/test_terminalbench.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        traj_file = tmp_path / "trajectory.json"
+        traj_file.write_text(json.dumps(trajectory))
+
+        # Parse the trajectory
+        with open(traj_file) as f:
+            parsed = json.load(f)
+
+        assert parsed["session_id"] == "test-session-123"
+        assert len(parsed["steps"]) == 2
+        assert parsed["final_metrics"]["total_cost_usd"] == 0.005
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/test_terminalbench.py]: Fixed in 5d4fed6 - added 6 new tests for convert_harbor_to_eval_output covering trajectory parsing, error handling, empty directories, and discovery logic.

File: benchmarks/terminalbench/config.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+# Default evaluation settings
+EVAL_DEFAULTS = {
+    "dataset": "terminal-bench-2",
+    "split": "test",
+}
+
+# Harbor configuration defaults
+HARBOR_DEFAULTS = {
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/config.py:11]: Fixed in a3e2d48: Now uses harbor_executable from config and extracts instruction from first user step.

File: benchmarks/terminalbench/run_infer.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+                "error": str(e),
+                "test_result": {},
+            })
+
+    # Fail if no trajectories were successfully processed
+    if not results and not errors:
+        raise RuntimeError(
+            f"No trajectories processed from {harbor_output_dir}"
+        )
+
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/run_infer.py:248]: Fixed in a3e2d48: Now uses harbor_executable from config and extracts instruction from first user step.

File: benchmarks/terminalbench/run_infer.py [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+        "--dataset",
+        type=str,
+        default=INFER_DEFAULTS["dataset"],
+        help="Harbor dataset name (e.g., terminal-bench@head, terminal-bench@2.0)",
+    )
+    parser.add_argument(
+        "--output-dir",
+        type=str,
+        default=INFER_DEFAULTS["output_dir"],
+        help="Base output directory for evaluation results",
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/run_infer.py:309]: Fixed in 5c3a1bb - Added documentation explaining that error entries use directory name (since we cannot parse the JSON to get session_id) while success entries use session_id from JSON. This edge case is now explicitly documented in the code comments.

File: benchmarks/terminalbench/run_infer.py [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+                "test_result": {
+                    "trajectory_path": str(traj_file),
+                    "total_steps": len(steps),
+                    "final_metrics": final_metrics,
+                },
+                "instruction": instruction,
+                "error": None,
+                "history": history,
+                "metrics": {
+                    "total_prompt_tokens": final_metrics.get("total_prompt_tokens", 0),
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/run_infer.py]: Fixed in 5c3a1bb - Now properly distinguishes between empty trials directory (no trial subdirectories found) vs missing trajectory files (trial directories exist but are missing trajectory.json files). Error messages are now more specific and actionable.

File: benchmarks/terminalbench/eval_infer.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+                        continue
+
+                    data = json.loads(line)
+                    instance_id = data.get("instance_id")
+                    if not instance_id:
+                        continue
+                    if instance_id in completed_ids or instance_id in incomplete_ids:
+                        continue
+
+                    incomplete_ids.add(instance_id)
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/eval_infer.py:165]: Fixed in 4239091 - Added documentation explaining this code exists for manual workflows where errors are extracted to a separate file for analysis.

File: benchmarks/terminalbench/config.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    "max_attempts": 1,
+    "max_retries": 3,
+    "workspace": "docker",  # docker or remote
+}
+
+# Default evaluation settings
+EVAL_DEFAULTS = {
+    "dataset": "terminal-bench-2",
+    "split": "test",
+}
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/config.py:8]: Fixed in 6431d1d - Removed EVAL_DEFAULTS and unused INFER_DEFAULTS entries. Added comment explaining error file handling.

File: benchmarks/terminalbench/config.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+INFER_DEFAULTS = {
+    "dataset": "terminal-bench-2",
+    "split": "test",
+    "output_dir": "./evaluation_outputs",
+    "n_limit": None,  # No limit by default
+    "num_workers": 1,
+    "max_iterations": 100,
+    "max_attempts": 1,
+    "max_retries": 3,
+    "workspace": "docker",  # docker or remote
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/config.py]: Fixed in 6431d1d - Removed EVAL_DEFAULTS and unused INFER_DEFAULTS entries. Added comment explaining error file handling.

File: benchmarks/terminalbench/eval_infer.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+                total_prompt_tokens += prompt_tokens or 0
+                total_completion_tokens += completion_tokens or 0
+
+            except json.JSONDecodeError as e:
+                logger.error(f"Line {line_num}: Invalid JSON - {e}")
+            except Exception as e:
+                logger.error(f"Line {line_num}: Unexpected error - {e}")
+
+    # Check for error file
+    error_path = Path(input_file).with_name(f"{Path(input_file).stem}_errors.jsonl")
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/eval_infer.py:149]: Fixed in 6431d1d - Removed EVAL_DEFAULTS and unused INFER_DEFAULTS entries. Added comment explaining error file handling.

**Comment by neubig** [comment]
@openhands run the full harbor eval run on one eval instance (make sure to start dockerd first), and also resolve all review comments using the github sparql API

File: benchmarks/terminalbench/eval_infer.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+                metrics = data.get("metrics", {})
+                final_metrics = test_result.get("final_metrics", {})
+
+                cost = metrics.get("total_cost_usd") or final_metrics.get(
+                    "total_cost_usd", 0.0
+                )
+                prompt_tokens = metrics.get("total_prompt_tokens") or final_metrics.get(
+                    "total_prompt_tokens", 0
+                )
+                completion_tokens = metrics.get(
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/eval_infer.py]: Fixed in f8beb28 - now uses explicit None check instead of `or` fallback to correctly handle zero values for cost and tokens.

File: benchmarks/terminalbench/run_infer.py [hunk 15]
```diff
@@ -0,0 +1,446 @@
...
+
+    # Add specific task IDs if provided
+    if task_ids:
+        for task_id in task_ids:
+            cmd.extend(["--task-id", task_id])
+
+    # Set up environment with LLM credentials
+    env = os.environ.copy()
+    if llm.api_key:
+        # api_key can be str or SecretStr
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/run_infer.py]: Fixed in f8beb28 - moved SecretStr import to top of file with other imports per project guidelines.

File: benchmarks/terminalbench/eval_infer.py [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+
+    try:
+        # Process results and generate report
+        process_terminalbench_results(
+            str(input_file),
+            str(output_file),
+        )
+
+        # Update Laminar datapoints with evaluation scores
+        LaminarService.get().update_evaluation_scores(str(input_file), str(output_file))
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/eval_infer.py:268]: Fixed in 2ae881d - wrapped both LaminarService.update_evaluation_scores() and generate_cost_report() in try/except blocks with warning-level logging to ensure multi-hour evaluations complete successfully even if these non-critical telemetry utilities fail.

File: benchmarks/terminalbench/eval_infer.py [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+                prompt_tokens = metrics.get("total_prompt_tokens")
+                if prompt_tokens is None:
+                    prompt_tokens = final_metrics.get("total_prompt_tokens", 0)
+
+                completion_tokens = metrics.get("total_completion_tokens")
+                if completion_tokens is None:
+                    completion_tokens = final_metrics.get("total_completion_tokens", 0)
+
+                total_cost_usd += cost if cost is not None else 0.0
+                total_prompt_tokens += prompt_tokens if prompt_tokens is not None else 0
```
[the remark this answers is not in the document]
Comment (neubig) [reply benchmarks/terminalbench/eval_infer.py]: Acknowledged - keeping the defensive ternary operators since they add clarity without hurting readability or performance, and provide an extra safety net given the recent zero-value bug fix.

File: tests/test_terminalbench.py [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+        output_file = tmp_path / "output.jsonl"
+
+        convert_harbor_to_eval_output(harbor_dir, output_file)
+
+        with open(output_file) as f:
+            entries = [json.loads(line) for line in f]
+
+        assert len(entries) == 5
+        instance_ids = {e["instance_id"] for e in entries}
+        assert instance_ids == {f"task-{i}" for i in range(5)}
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/test_terminalbench.py:399]: Thanks for the positive feedback! The tests are designed to validate real behavior - trajectory parsing, error aggregation, and edge cases - rather than just asserting mocked calls.

**Comment by neubig** [comment]
I confirmed that this works in my local terminal, merging in.
